### PR TITLE
Karma: enable coverage reports

### PR DESCRIPTION
**Pull request by dvoytenko**
## Summary

Enable coverage report for Karma tests based on the [coverage-istanbul-loader](https://github.com/JS-DevTools/coverage-istanbul-loader/tree/master/examples/karma).

## Relevant Technical Choices

As you can see the coverage is reported to up ~20%. We certainly need to track the coverage deltas we are getting through Karma, but the question is: Is this a good number in absolute terms at this point? Some (big?) percent of that added coverage we get from the rendering phase for the whole editor fixture. And while we are getting some assurance that the rendering is not failing, it's also not the true coverage either. This number will be a little more justifiable if we add more quality control tools to the Karma fixture, such as:
 1. Accessibility tests.
 2. [✅] Visual tests (Percy, which we are pretty close too).
 3. Some performance metrics such as rerendering volume.
 4. If we pursue aggressively React-Dev warnings.

But as it standard right now, the relative value is definitely useful to us, but the absolute value is not directly comparable to the existing coverage from unit tests.

Another nuance: I tried adding the `flags` field to our coverage reports and you can see the breakdown below. That definitely helps, but the analytics by flag is currently limited. [This](https://user-images.githubusercontent.com/726049/84980016-ca85ff00-b0e5-11ea-87f8-80a8167311ce.png) is what you'd see if you'd try to drill down.

## To-do

- [x] Discuss the tradeoffs
- [x] Separate Karma coverage into a separate project

---

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -132,6 +132,8 @@
     "jsdom-screenshot": "^3.2.0",
     "karma": "^5.1.0",
     "karma-chrome-launcher": "^3.1.0",
+    "karma-coverage-istanbul-reporter": "^3.0.3",
+    "@jsdevtools/coverage-istanbul-loader": "^3.0.3",
```
Comment (swissspidy) [inline package.json]: Packages should be listed alphabetically in this file

File: webpack.config.test.cjs [hunk 2]
```diff
@@ -29,22 +29,35 @@ const DependencyExtractionWebpackPlugin = require('@wordpress/dependency-extract
...
+      // otherwise has some failures.
+      plugins: webpackConfig.plugins.filter(
+        (plugin) => !(plugin instanceof DependencyExtractionWebpackPlugin)
+      ),
+    }))[0];
+  if (coverage) {
+    config.module.rules.push({
+      enforce: 'post',
+      test: /\.js$/,
+      exclude: [/karma/, /\.karma\.js$/, /\.test\.js$/, /node_modules/],
```
Comment (swissspidy) [inline webpack.config.test.cjs]: Maybe make this less broad, and explicitly target `karma` folders  & files with `.karma.js` files?

```suggestion
      exclude: [/\/karma\//, /\.karma\.js$/, /\.test\.js$/, /node_modules/],
```

File: webpack.config.test.cjs [hunk 3]
```diff
@@ -29,22 +29,35 @@ const DependencyExtractionWebpackPlugin = require('@wordpress/dependency-extract
...
-      ...webpackConfig.output,
-      path: path.resolve(process.cwd(), 'assets', 'testjs'),
-    },
-    // WP's DependencyExtractionWebpackPlugin is not needed for tests and
-    // otherwise has some failures.
-    plugins: webpackConfig.plugins.filter(
-      (plugin) => !(plugin instanceof DependencyExtractionWebpackPlugin)
-    ),
-  }))[0];
+function getConfig(group, { coverage = false } = {}) {
```
Comment (swissspidy) [inline webpack.config.test.cjs:32]: 👍 for passing the group like this

File: package.json (hunk 1 above)
Comment (dvoytenko) [reply package.json]: Prettier doesn't do that?

Comment (swissspidy) [reply package.json]: `npm` does this usually automatically when doing `npm install` or similar. Prettier thus leaves that alone I think

**Comment by dvoytenko** [comment]
🎉 After a few attempts I tried the most obvious thing and it worked. Now `unit` and `karma` tests are split in subprojects and everything looks pretty solid. Please rereview and let's get this in!

**Comment by dvoytenko** [comment]
For posterity, here's how the coverage numbers change:

_                             | Before   | After    
-----------------|--------|--------
codecov/project   | 66.44% | 77.94% 
codecov/project/editor   | 66.1% | 83.93%
codecov/project/editor.unit   | 66.1% | 66.1%
codecov/project/editor.karma   | - | 62.47%

**Review by swissspidy — APPROVED** [review]
Really nice! 💪 🎉

**Pull request merged**